src/store/operations/tools.ts

```diff
@@ -0,0 +1,95 @@
+import { AnnotationToolStore } from '@/src/store/tools/useAnnotationTool';
+import { Maybe } from '@/src/types';
+import { AnnotationTool } from '@/src/types/annotation-tool';
+import { IHistoryOperation } from '@/src/types/history';
+import { Store } from 'pinia';
+
+export function createAddToolOperation<
+  ID extends string,
+  T extends AnnotationTool<ID>,
+  S extends AnnotationToolStore<ID> & Store
+>(store: S, initToolState: Partial<T>): IHistoryOperation<ID> {
+  let id: Maybe<ID> = null;
+  let toolState: Maybe<T> = null;
+
+  const isApplied = () => id != null;
+
+  const apply = () => {
+    if (isApplied()) return id!;
+    id = store.addTool(toolState ?? initToolState);
+    return id;
+  };
+
+  const revert = () => {
+    if (!isApplied()) return;
+    toolState = store.toolByID[id!] as T;
+    store.removeTool(id!);
+    id = null;
+  };
+
+  return {
+    apply,
+    revert,
+    isApplied,
+  };
+}
+
+export function createRemoveToolOperation<
+  ID extends string,
+  S extends AnnotationToolStore<ID> & Store
+>(store: S, id: ID): IHistoryOperation {
+  let lastID: ID = id;
+  let tool: Maybe<AnnotationTool<ID>> = null;
+
+  const isApplied = () => tool != null;
+
+  const apply = () => {
+    if (isApplied()) return;
+    tool = store.toolByID[lastID];
+    store.removeTool(lastID);
+  };
+
+  const revert = () => {
+    if (!isApplied()) return;
+    lastID = store.addTool(tool!);
+    tool = null;
+  };
+
+  return {
+    apply,
+    revert,
+    isApplied,
+  };
+}
+
+export function createUpdateToolOperation<
```

Comment:
```suggestion
// Not used at the moment
export function createUpdateToolOperation<
```

Reply:
I intended to add support for undoing stuff like changing labels, moving rectangle points, etc. That hasn't happened yet, so I'm good to omit this function for now instead of adding a comment that we will have to later remove anyways.